src/NHibernate/Util/SoftLimitMRUCache.cs

```diff
@@ -22,33 +22,33 @@ namespace NHibernate.Util
 	/// enqueuement.
 	/// </remarks>
 	[Serializable]
-	public class SoftLimitMRUCache : IDeserializationCallback
+	public class SoftLimitMRUCache<TKey, TValue> : IDeserializationCallback where TKey : class where TValue : class
```

Comment:
As said on first commit review, I question whether it is worth it or not. On Java side, they have ceased using this for query plan caching. (Java side, they have start using that weak hashtable for first level cache it seems.) They are using a 2048 max entries (configurable) plan cache instead with an algorithm for evicting coldest entries. And moreover, they use something more efficient than weak references for their remaining "caching" purpose. (Usage of weak reference for caching is by the way advises against by Microsoft documentation.) Weak references may be reclaimed while being far from reaching available memory limits. Java use soft references instead, designed to be reclaimed only when under memory stress. They do not have equivalent in .Net. 

We may have better throwing that away for using a `System.Runtime.Caching.MemoryCache` instead (available since fx4).